### PR TITLE
T3485: Add macos support

### PR DESCRIPTION
**Pull request by rchenzheng**
## Change Summary

Allows building of images in macOS through vagrant/virtualbox/docker

  - add binary dependency checks
  - add os check
  - automate image building
  - use rsync to solely transfer image to local

## Types of changes

 
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)

https://phabricator.vyos.net/T3485

## Component(s) name

vyos-build

## Proposed changes

Create an iso in MacOS in an efficient manner using vagrant, virtualbox and docker which tasks are embedded in the make task `macos-build`

## How to test

`make macos-build`

## Checklist:

- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [x] My commit headlines contain a valid Task id
- [x] My change requires a change to the documentation
- [x] I have updated the documentation accordingly

### CODE REVIEW

**Review by dmbaturin — CHANGES_REQUESTED** [review]
I'm not going to deny that the use case exists. However, I believe this should be an external frontend script rather than a built-in, for multiple reasons:

* None of the maintainers are building on OS X, so we will not be immediately alarmed when this code breaks. For us, it's a liability, not an asset. Code should be maintained by people who have a use for it even if just for this reason.
* It's not exactly "macOS support", but macOS+VirtualBox+Vagrant+etc. It's not a bad stack by itself, but it's a lot of moving parts that some people may or may not want.
* There are quite a lot of hardcoded, non-configurable bits too, from the internal network to the default build-by field. That's a functionality loss compared to the real build scripts.

In the current shape, I'd happily merge that code into https://github.com/vyos/vyos-utils-misc as an "at your own risk" piece. I can't see it belonging in the vyos-build repository.

**Comment by rchenzheng** [comment]
> I'm not going to deny that the use case exists. However, I believe this should be an external frontend script rather than a built-in, for multiple reasons:
> 
> * None of the maintainers are building on OS X, so we will not be immediately alarmed when this code breaks. For us, it's a liability, not an asset. Code should be maintained by people who have a use for it even if just for this reason.
> * It's not exactly "macOS support", but macOS+VirtualBox+Vagrant+etc. It's not a bad stack by itself, but it's a lot of moving parts that some people may or may not want.
> * There are quite a lot of hardcoded, non-configurable bits too, from the internal network to the default build-by field. That's a functionality loss compared to the real build scripts.
> 
> In the current shape, I'd happily merge that code into https://github.com/vyos/vyos-utils-misc as an "at your own risk" piece. I can't see it belonging in the vyos-build repository.

- I believe there is currently a lot of steps just to build it outside of debian never mind the two versions, but I believe this should be somewhere so everyone can use VyOS. In the current state, it's rather cumbersome to develop in MAC which most people do. My goal was to make it easier to contribute to VyOS

- As for hardcoding, the only thing hardcoded but fictitious is the internal IP and the branch hence why there are 2 pull requests.

How do you prefer to merge this onto the other repo?

**Comment by jack9603301** [comment]
@rchenzheng If Docker is available for Mac, I don't think we need additional build scripts, since the compilation is always done in Debian Docker

**Comment by rchenzheng** [comment]
> @rchenzheng If Docker is available for Mac, I don't think we need additional build scripts, since the compilation is always done in Debian Docker

Unfortunately, there are devices that are only available in Debian. So the only way is to spin up a Debian VM.

**Comment by dmbaturin** [comment]
> How do you prefer to merge this onto the other repo?

Extract the logic into a separate script and make a new pull request against vyos-utils-misc.

**Comment by jack9603301** [comment]
If this is the case, Mac developers can often easily launch the virtual machine, just as I used the virtual machine for related testing. Therefore, there is no need for related scripts, and this can be done from the following perspective:

1. Convert the docker image into a virtual machine image iso, or build a Build VM specifically for vyos, and then users can install this system in any virtual machine management system, and develop on this system (I think it only needs to be in debian Write an initial vyos build script, and then distribute it)

File: scripts/check-macos-build-env [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+else:
+    checker.print_missing_deps()
+    if 'vagrant' in missing['binaries']:
+        print("Your system does not have Vagrant.")
+        print("Please install Vagrant from https://www.vagrantup.com/downloads.")
+        sys.exit(1)
+    elif 'virtualbox' in missing['binaries']:
+        print("Your system does not have Virtualbox.")
+        print("Please install Virtualbox from https://www.virtualbox.org/wiki/Downloads.")
+        sys.exit(1)
```
Comment (jack9603301) [inline scripts/check-macos-build-env:60]: This will require the user to use the specified hypervisor, which is not appropriate

**Review by jack9603301 — CHANGES_REQUESTED** [review]

File: scripts/check-macos-build-env (hunk 1 above)
Comment (rchenzheng) [reply scripts/check-macos-build-env:60]: It's the only free option available and compatible with vagrant

**Comment by rchenzheng** [comment]
> If this is the case, Mac developers can often easily launch the virtual machine, just as I used the virtual machine for related testing. Therefore, there is no need for related scripts, and this can be done from the following perspective:
> 
> 1. Convert the docker image into a virtual machine image iso, or build a Build VM specifically for vyos, and then users can install this system in any virtual machine management system, and develop on this system (I think it only needs to be in debian Write an initial vyos build script, and then distribute it)

It's time consuming having to build a VM, by the time someone's done building a VM vagrant has already spit out the image.

I'll create a PR on the other repo. shorly

File: scripts/check-macos-build-env (hunk 1 above)
Comment (jack9603301) [reply scripts/check-macos-build-env:60]: If a developer has already installed other virtual machines, they must also install a new VM for building vyos? I suggest that this is a suggested option, not a seemingly mandatory one

Comment (rchenzheng) [reply scripts/check-macos-build-env:60]: From my perspective, you may run into but "it works in my machine" scenario as the VM itself might be a snowflake.

They both work but using vagrant is easier.

Comment (jack9603301) [reply scripts/check-macos-build-env:60]: Ok i see what you mean, thank you

**Comment by rchenzheng** [comment]
Moved to https://github.com/vyos/vyos-utils-misc/pull/1

**Pull request closed without merging**